src/clib/map.d

```diff
@@ -143,12 +366,595 @@ struct map(K, V, A: IAllocator!V = allocator!V) {
+            }
+        }
+        return count;
     }
 
     private Node* alloc_node() @nogc nothrow {
         if (_allocator is null) _allocator = _new!A();
-        return cast(Node*) _allocator.allocate_vptr(Node.sizeof);
+        Node* node = cast(Node*) _allocator.allocate_vptr(Node.sizeof);
+        if (node !is null) {
+            node.filled = 0;
+            node.is_leaf = true;
+            node.parent = null;
+            foreach(i; 0..Node.N) {
+                node.children[i] = null;
+            }
+        }
+        return node;
+    }
+
+    // Iterator structs
+    private struct MapIterator(K, V) {
+        map!(K, V)* container;
+        Node* current;
+        size_t index;
+        bool isEnd;
+
+        this(map!(K, V)* c, bool end) @nogc nothrow {
+            container = c;
+            isEnd = end;
+            if (!end && container.root !is null) {
+                current = container.root;
+                while (current.children[0] !is null) {
+                    current = current.children[0];
+                }
+                index = 0;
+            } else {
+                current = null;
+                index = 0;
+            }
+        }
+
+        ref TreePair!(K, V) front() @nogc nothrow {
+            assert(!empty, "Attempting to fetch the front of an empty range");
+            return current.pairs[index];
+        }
+
+        ref MapIterator popFront() @nogc nothrow {
+            assert(!empty, "Attempting to popFront an empty range");
+            if (index + 1 < current.filled) {
+                index++;
+            } else {
+                const(Node)* next = container.next_inorder_node(current);
+                if (next is null) {
+                    isEnd = true;
+                    current = null;
+                } else {
+                    current = cast(Node*)next;
+                    index = 0;
+                }
+            }
+            return this;
+        }
+
+        bool empty() const @nogc nothrow {
+            return isEnd || current is null || container.root is null;
+        }
+
+        @property MapIterator save() @nogc nothrow {
+            return this;
+        }
+    }
+
+    private struct ConstMapIterator(K, V) {
+        const map!(K, V)* container;
+        const(Node)* current;
+        size_t index;
+        bool isEnd;
+
+        this(const map!(K, V)* c, bool end) @nogc nothrow {
+            container = c;
+            isEnd = end;
+            if (!end && container.root !is null) {
+                current = container.root;
+                while (current.children[0] !is null) {
+                    current = current.children[0];
+                }
+                index = 0;
+            } else {
+                current = null;
+                index = 0;
+            }
+        }
+
+        ref const(TreePair!(K, V)) front() const @nogc nothrow {
+            assert(!empty, "Attempting to fetch the front of an empty range");
```

Comment:
same
just a note, i think ive did something similar for error handling in list, gonna have to check